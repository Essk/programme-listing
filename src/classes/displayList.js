/**
 * A Programme
 * @typedef {Object} Programme
 * @property {string} name - The name
 * @property {string} description - The description
 * @property {boolean} active - whether active or not
 */

/**
 * A Programme with Id
 * @typedef {Object} ProgrammeWithId
 * @property {number} id - unique id, should be integer
 * @property {string} name - The name
 * @property {string} description - The description
 * @property {boolean} active - whether active or not
 */
export class DisplayList {
  constructor(programmes) {
    this.programmeList = programmes
      ? this.cleanUpInput(programmes).map(p => {
          p.show = true;
          return p;
        })
      : [];
  }
  /**
   *
   * @param {object} rawInput needs to have a results array
   * @returns {ProgrammeWithId[]}
   */
  cleanUpInput(rawInput) {
    return rawInput.results.map(programme => {
      return {
        id: programme.id,
        name: programme.name,
        description: programme.shortDescription,
        active: programme.active,
      };
    });
  }
  /**
   * stubbing the fact that we'd probably have an external service checking this against a DB in a real app
   *
   * @returns {number}
   */
  fetchNewId() {
    const existingIDs = this.programmeList.map(programme =>
      parseInt(programme.id)
    );
    return existingIDs[existingIDs.length - 1] + 1;
  }

  /**
   * @param {number} index Array index
   * @return {ProgrammeWithId | object} a prgramme object or an empty object
   */
  getProgrammeByIndex(index) {
    return this.programmeList[index] || {};
  }

  /**
   * @param {number | string} id id of programme - strings will be converted to integers
   * @return {ProgrammeWithId | object} a prgramme object or an empty object
   */
  getProgrammeById(id) {
    const parsedId = parseInt(id);
    return (
      this.programmeList.find(programme => programme.id === parsedId) || {}
    );
  }

  /**
   * @param {Programme} newProgramme to be added
   * @return {number} id of the programme
   */
  addProgramme(newProgramme) {
    const newId = this.fetchNewId();
    newProgramme.id = newId;
    this.programmeList.push(newProgramme);
    return this;
  }

  /**
   * @param {number} id of programme to be deleted
   * @return {ProgrammeWithId || object} programme object that was deleted or empty object
   */
  deleteProgramme(id) {
    const idx = this.programmeList.findIndex(
      programme => programme.id === parseInt(id)
    );
    if (idx === -1) {
      return {};
    }
    this.programmeList.splice(idx, 1);
    return this;
  }
  /**
   *
   * @param {number|string} id
   * @param {Programme} programme
   * @returns {ProgrammeWithId|object}
   */
  updateProgramme(id, programme) {
    const idx = this.programmeList.findIndex(
      programme => programme.id === parseInt(id)
    );
    if (idx === -1) {
      return {};
    }
    programme.id = id;
    this.programmeList.splice(idx, 1, programme);
    return this;
  }
  search(str) {
    this.programmeList = this.programmeList.map(programme => {
      programme.show = programme.name.toLowerCase().includes(str.toLowerCase());
      return programme;
    });
    return this;
  }
}
