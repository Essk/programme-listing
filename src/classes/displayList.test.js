import programmes from '../programmes.json';
import { DisplayList } from '../classes/displayList';

describe('DisplayList', () => {
  it('instantiates', () => {
    const displayList = new DisplayList(programmes);
    expect(displayList).toBeTruthy;
  });
  it('contains a list of programmes', () => {
    const displayList = new DisplayList(programmes);
    expect(displayList.programmeList).toMatchSnapshot();
  });
  describe('cleanUpInput', () => {
    const displayList = new DisplayList();
    it('extracts the results portion of incomming data object', () => {
      const rawData = {
        stuff: 'things',
        results: [
          {
            property: 'value',
            id: 'id',
            name: 'name',
            shortDescription: 'shortDescription',
            active: 'active',
          },
        ],
      };
      const results = displayList.cleanUpInput(rawData);
      expect(results.length).toBe(rawData.results.length);
      results.forEach(result => {
        expect(result).toHaveProperty('id');
        expect(result).toHaveProperty('name');
        expect(result).toHaveProperty('description');
        expect(result).toHaveProperty('active');
      });
    });
  });
  it.todo('can add a programme');
  it.todo('can delete a programme');
  it.todo('can update a prgramme');
  it.todo('can return a programme by id');
  it.todo('can return a programme by index');
});
