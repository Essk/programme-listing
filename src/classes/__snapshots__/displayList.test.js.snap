// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DisplayList contains a list of programmes 1`] = `
Array [
  Object {
    "active": true,
    "description": "Another look at some of the nation's oldest drivers, including 102-year-old Harry Kartz.",
    "id": 1938,
    "name": "100 Year Old Drivers Rebooted",
  },
  Object {
    "active": false,
    "description": "Quiz show in which contestants have their heart rate constantly monitored.",
    "id": 1595,
    "name": "1000 Heartbeats",
  },
  Object {
    "active": true,
    "description": "Documentary celebrating the 50th anniversary of England hosting and winning the World Cup.",
    "id": 1951,
    "name": "1966 - A Nation Remembers",
  },
  Object {
    "active": false,
    "description": "Documentary celebrating Taggart's 30th anniversary.",
    "id": 1262,
    "name": "30 Years of Taggart on TV",
  },
  Object {
    "active": false,
    "description": "Bradley Walsh hosts a festive spectacular with a variety of guests.",
    "id": 1553,
    "name": "A Christmas Cracker",
  },
  Object {
    "active": false,
    "description": "Jason Manford presents this end-of-year comedy review show.",
    "id": 1563,
    "name": "A Funny Old Year",
  },
  Object {
    "active": true,
    "description": "Griff Rhys Jones explores his native land.",
    "id": 1676,
    "name": "A Great Welsh Adventure with Griff Rhys Jones",
  },
  Object {
    "active": false,
    "description": "Selena Scott explores the challenges facing the  the National Trust.",
    "id": 1439,
    "name": "A Matter of Trust",
  },
  Object {
    "active": false,
    "description": "Phillip Schofield and Amanda Holden host the annual military awards ceremony.",
    "id": 336,
    "name": "A Night of Heroes: The Sun Military Awards",
  },
  Object {
    "active": false,
    "description": "Story of a real-life Scottish criminal who spent a decade in and out of British prisons.",
    "id": 1860,
    "name": "A Sense of Freedom",
  },
  Object {
    "active": false,
    "description": "Crime drama starring David Jason.",
    "id": 207,
    "name": "A Touch of Frost",
  },
  Object {
    "active": true,
    "description": "The issue of child abduction and three parents struggling to get their children back.",
    "id": 1707,
    "name": "Abducted",
  },
  Object {
    "active": true,
    "description": "New series in which Adrian Edmondson explores Britain's maritime past and its influence.",
    "id": 1373,
    "name": "Ade At Sea",
  },
  Object {
    "active": false,
    "description": "Adrian Edmondson travels round Britain in search of foods and traditions.",
    "id": 523,
    "name": "Ade in Britain",
  },
  Object {
    "active": true,
    "description": "After Midnight",
    "id": 1767,
    "name": "After Midnight",
  },
  Object {
    "active": false,
    "description": "After They Were Famous",
    "id": 557,
    "name": "After They Were Famous",
  },
  Object {
    "active": false,
    "description": "Series based on Agatha Christie's amateur sleuth, with Julia McKenzie.",
    "id": 154,
    "name": "Agatha Christie's Marple",
  },
  Object {
    "active": false,
    "description": "Drama featuring Agatha Christie's Belgian detective, Hercule Poirot.",
    "id": 476,
    "name": "Agatha Christie's Poirot",
  },
  Object {
    "active": true,
    "description": "Robson Green and Jerome Flynn star in this wartime comedy drama.",
    "id": 1558,
    "name": "Ain't Misbehavin'",
  },
  Object {
    "active": true,
    "description": "Alexander Armstrong journeys halfway around the Arctic Circle.",
    "id": 1760,
    "name": "Alexander Armstrong in the Land of the Midnight Sun",
  },
  Object {
    "active": true,
    "description": "Vernon Kay presents the celebrity family quiz show.",
    "id": 90,
    "name": "All Star Family Fortunes",
  },
  Object {
    "active": true,
    "description": "Three celebrity couples try to win cash for their chosen charity.",
    "id": 2,
    "name": "All Star Mr & Mrs",
  },
  Object {
    "active": false,
    "description": "Dance comedy starring Theo Stevenson and Akai Osei-Mansfield.",
    "id": 1551,
    "name": "All Stars",
  },
  Object {
    "active": true,
    "description": "Along Came Polly",
    "id": 1720,
    "name": "Along Came Polly",
  },
  Object {
    "active": false,
    "description": "Paddy McGuinness and Angela Rippon present as young challengers take on senior citizens.",
    "id": 1389,
    "name": "Amazing Greys",
  },
  Object {
    "active": true,
    "description": "Anatole is a charming but excitable French mouse who lives under the streets of Paris.",
    "id": 1599,
    "name": "Anatole",
  },
  Object {
    "active": false,
    "description": "Britain’s newsreaders reveal what life is like behind the desk, and how it has changed.",
    "id": 1745,
    "name": "And Here is the News...",
  },
  Object {
    "active": true,
    "description": "Scotland's leading animal charity allows access to the work of its inspectors.",
    "id": 597,
    "name": "Animal 999",
  },
  Object {
    "active": false,
    "description": "Series about the Royal Army Veterinary Corps and the bond between soldiers and animals.",
    "id": 1256,
    "name": "Animal Heroes",
  },
  Object {
    "active": true,
    "description": "Enthralling natural world series about the challenge of motherhood in the wild.",
    "id": 1717,
    "name": "Animal Mums",
  },
  Object {
    "active": false,
    "description": "Ant and Dec return with their live family entertainment show.",
    "id": 1152,
    "name": "Ant & Dec's Saturday Night Takeaway",
  },
  Object {
    "active": false,
    "description": "Drama based on the relationship between Sir Arthur Conan Doyle and George Edalji.",
    "id": 1607,
    "name": "Arthur & George",
  },
  Object {
    "active": false,
    "description": "NULL",
    "id": 512,
    "name": "Arts & Parts",
  },
  Object {
    "active": false,
    "description": "Series following pet owners who are forced to find new homes for their much-loved animals.",
    "id": 1371,
    "name": "Auf Wiedersehen My Pet",
  },
  Object {
    "active": false,
    "description": "Dominic Littlewood confronts dodgy builders and meets people whose lives they have ruined.",
    "id": 1586,
    "name": "Bad Builders Bang to Rights",
  },
  Object {
    "active": false,
    "description": "BAFTA honours Downton Abbey with a special award presented by Julie Walters.",
    "id": 1795,
    "name": "BAFTA Celebrates Downton Abbey",
  },
  Object {
    "active": true,
    "description": "Rickie Haywood Williams and Melvin Odoom host as two teams of friends battle it out.",
    "id": 1879,
    "name": "Bang on the Money",
  },
  Object {
    "active": true,
    "description": "A look at the changing face of UK retail - Costco, Motorpoint, Shoe Zone & Approved Food.",
    "id": 1642,
    "name": "Bargain Fever Britain",
  },
  Object {
    "active": true,
    "description": "Series going behind the scenes of Britain's newest bargain chain Pep and Co.",
    "id": 1877,
    "name": "Bargain Shop Wars",
  },
  Object {
    "active": true,
    "description": "Series in which John Sergeant sets off along Britain's eight best canal journeys.",
    "id": 1869,
    "name": "Barging Round Britain with John Sergeant",
  },
  Object {
    "active": true,
    "description": "Competition to find the UK's best BBQer with judges Adam Richman and Mark Blatchford.",
    "id": 1697,
    "name": "BBQ Champ",
  },
  Object {
    "active": true,
    "description": "Eight celebrities take a South African bush journey to learn the skills to survive.",
    "id": 1594,
    "name": "Bear Grylls: Mission Survive",
  },
  Object {
    "active": false,
    "description": "Beat the Star",
    "id": 3,
    "name": "Beat the Star",
  },
  Object {
    "active": true,
    "description": "Documentary about one of cricket's most colourful and controversial players.",
    "id": 1788,
    "name": "Being Kevin Pietersen",
  },
  Object {
    "active": true,
    "description": "Comedy series set in Benidorm at the Solana all-inclusive resort.",
    "id": 359,
    "name": "Benidorm",
  },
  Object {
    "active": true,
    "description": "An epic re-imagining of one of literature’s greatest and most enduring heroes, Beowulf.",
    "id": 1790,
    "name": "Beowulf: Return to the Shieldlands",
  },
  Object {
    "active": true,
    "description": "A look back at the best bits of local telly from the last 12 months.",
    "id": 1816,
    "name": "Best of the Year 2015",
  },
  Object {
    "active": true,
    "description": "Series following author and broadcaster Tom Weir's travels in Scotland.",
    "id": 1079,
    "name": "Best of Tom Weir",
  },
  Object {
    "active": true,
    "description": "Julia Bradbury takes us along on some of her favourite walks with the very best views.",
    "id": 1838,
    "name": "Best Walks with a View with Julia Bradbury",
  },
  Object {
    "active": true,
    "description": "Five households test the latest gadgets to find out if they are worth the money.",
    "id": 1841,
    "name": "Big Box Little Box",
  },
]
`;
